pandas/core/frame.py

```diff
@@ -5219,7 +5219,21 @@ def _sanitize_column(self, value) -> tuple[ArrayLike, BlockValuesRefs | None]:
 
         if is_list_like(value):
             com.require_length_match(value, self.index)
-        return sanitize_array(value, self.index, copy=True, allow_2d=True), None
+        arr = sanitize_array(value, self.index, copy=True, allow_2d=True)
+        if (
+            isinstance(value, Index)
```

Comment:
might be more performant to do the Index check before the sanitize_array?

Reply:
Why? We have to sanitise anyway?

Reply:
if you know you have an Index (and no dtype), sanitize_array boils down to:

```
if len(value) != len(index): raise ...
return value._values
```

so a lot of sanitize_array may be made unnecessary in this case

Reply:
That's what I want in the future, but we go through ``maybe_infer_to_datetimelike`` at the moment which changes the dtype, so we can't shortcut this before we enforce the deprecation